ariadne_codegen/contrib/client_forward_refs.py

```diff
@@ -0,0 +1,375 @@
+            call = self._get_call_arg_from_async_for(return_stmt)
+        else:
+            return
+
+        if call is None:
+            return
+
+        import_class = self._get_class_from_call(call)
+        if import_class is None:
+            return
+
+        import_class_id = import_class.id
+
+        # We add the class to our set of imported in methods - these classes
+        # don't need to be imported at all in the global scope.
+        self.imported_in_method.add(import_class.id)
+        method_def.body.insert(
+            0,
+            ast.ImportFrom(
+                module=self.imported_classes[import_class_id],
+                names=[import_class],
+            ),
+        )
+
+    def _get_call_arg_from_return(self, return_stmt: ast.Return) -> Optional[ast.Call]:
+        """Get the class used in the return statement.
+
+        :param return_stmt: The statement used for return
+        """
+        # If it's a call of the class like produced by
+        # `ShorterResultsPlugin` we have an attribute.
+        if isinstance(return_stmt.value, ast.Attribute) and isinstance(
+            return_stmt.value.value, ast.Call
+        ):
+            return return_stmt.value.value
+
+        # If not it's just a call statement to the generated class.
+        if isinstance(return_stmt.value, ast.Call):
+            return return_stmt.value
+
+        return None
+
+    def _get_call_arg_from_async_for(
+        self, last_stmt: ast.AsyncFor
+    ) -> Optional[ast.Call]:
+        """Get the class used in the yield expression.
+
+        :param last_stmt: The statement used in `ast.AsyncFor`
+        """
+        if isinstance(last_stmt.body, list) and isinstance(last_stmt.body[0], ast.Expr):
+            body = last_stmt.body[0]
+        elif isinstance(last_stmt.body, ast.Expr):
+            body = last_stmt.body
+        else:
+            return None
+
+        if not isinstance(body, ast.Expr):
+            return None
+
+        if not isinstance(body.value, ast.Yield):
+            return None
+
+        # If it's a call of the class like produced by
+        # `ShorterResultsPlugin` we have an attribute.
+        if isinstance(body.value.value, ast.Attribute) and isinstance(
+            body.value.value.value, ast.Call
+        ):
+            return body.value.value.value
+
+        # If not it's just a call statement to the generated class.
+        if isinstance(body.value.value, ast.Call):
+            return body.value.value
+
+        return None
+
+    def _get_class_from_call(self, call: ast.Call) -> Optional[ast.Name]:
+        """Get the class from an `ast.Call`.
+
+        :param call: The `ast.Call` arg
+        :returns: `ast.Name` or `None`
+        """
+        if not isinstance(call.func, ast.Attribute):
+            return None
+
+        if not isinstance(call.func.value, ast.Name):
+            return None
+
+        return call.func.value
+
+    def _update_imports(self, module: ast.Module):
```

Comment:
Did You consider splitting this function into a smaller ones? Instead of big comments in the middle of the function we could have a smaller functions with docstring that can reduce the complexity of this code. Those two `for` loops looks like they can be separated. Just a suggestion :D

Reply:
Are you referring to the comment about why we need to be careful wit the `ImportFrom` nodes? I didn't consider splitting it further since they both work on the imports but can absolutely do that. One loop removes the forward refs and the other one adds them.  It would be easy to split by that logic but I would keep the comment anyway since it adds valuable context to why we do it.

Reply:
Split it into two methods, moved some comments to docstrings